File: cdm-sources/src/cdm/margin/schedule/functions/GetGrossInitialMarginFromStandardizedSchedule.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.observable.asset.Money import Money
from cdm.margin.schedule.StandardizedSchedule import StandardizedSchedule
from cdm.margin.schedule.functions.GetStandardizedScheduleMarginRate import GetStandardizedScheduleMarginRate

__all__ = ['GetGrossInitialMarginFromStandardizedSchedule']


@replaceable
def GetGrossInitialMarginFromStandardizedSchedule(standardizedSchedule: StandardizedSchedule) -> Money:
    """
    Takes the grid information from an specific trade and calculates the gross initial margin.
    
    Parameters 
    ----------
    standardizedSchedule : StandardizedSchedule
    
    Returns
    -------
    grossInitialMargin : Money
    
    """
    _post_registry = {}
    self = inspect.currentframe()
    
    
    initialMarginRequirement = GetStandardizedScheduleMarginRate(rosetta_resolve_attr(rosetta_resolve_attr(self, "standardizedSchedule"), "assetClass"), rosetta_resolve_attr(rosetta_resolve_attr(self, "standardizedSchedule"), "durationInYears"))
    grossInitialMargin = _get_rosetta_object('Money', 'value', ((rosetta_resolve_attr(rosetta_resolve_attr(self, "standardizedSchedule"), "notional") * rosetta_resolve_attr(self, "initialMarginRequirement")) * 0.01))
    grossInitialMargin = set_rosetta_attr(rosetta_resolve_attr(self, 'grossInitialMargin'), 'unit->currency', rosetta_resolve_attr(rosetta_resolve_attr(self, "standardizedSchedule"), "notionalCurrency"))
    
    # post-conditions
    
    @rosetta_local_condition(_post_registry)
    def condition_0_PositiveGrossInitialMargin(self):
        """
        Ensure gross initial margin is greater than 0
        """
        return all_elements(rosetta_resolve_attr(rosetta_resolve_attr(self, "grossInitialMargin"), "value"), ">", 0)
    # Execute all registered post-conditions
    execute_local_conditions(_post_registry, 'Post-condition')
    
    return grossInitialMargin

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/margin/schedule/functions/GetGrossInitialMarginFromStandardizedSchedule.java
package cdm.margin.schedule.functions;

import cdm.margin.schedule.StandardizedSchedule;
import cdm.margin.schedule.StandardizedScheduleAssetClassEnum;
import cdm.observable.asset.Money;
import cdm.observable.asset.Money.MoneyBuilder;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.expression.MapperMaths;
import com.rosetta.model.lib.functions.ConditionValidator;
import com.rosetta.model.lib.functions.ModelObjectValidator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import java.math.BigDecimal;
import java.util.Optional;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(GetGrossInitialMarginFromStandardizedSchedule.GetGrossInitialMarginFromStandardizedScheduleDefault.class)
public abstract class GetGrossInitialMarginFromStandardizedSchedule implements RosettaFunction {
	
	@Inject protected ConditionValidator conditionValidator;
	
	@Inject protected ModelObjectValidator objectValidator;
	
	// RosettaFunction dependencies
	//
	@Inject protected GetStandardizedScheduleMarginRate getStandardizedScheduleMarginRate;

	/**
	* @param standardizedSchedule 
	* @return grossInitialMargin 
	*/
	public Money evaluate(StandardizedSchedule standardizedSchedule) {
		Money.MoneyBuilder grossInitialMarginBuilder = doEvaluate(standardizedSchedule);
		
		final Money grossInitialMargin;
		if (grossInitialMarginBuilder == null) {
			grossInitialMargin = null;
		} else {
			grossInitialMargin = grossInitialMarginBuilder.build();
			objectValidator.validate(Money.class, grossInitialMargin);
		}
		
		// post-conditions
		conditionValidator.validate(() -> greaterThan(MapperS.of(grossInitialMargin).<BigDecimal>map("getValue", money -> money.getValue()), MapperS.of(BigDecimal.valueOf(0)), CardinalityOperator.All),
			"Ensure gross initial margin is greater than 0");
		
		return grossInitialMargin;
	}

	protected abstract Money.MoneyBuilder doEvaluate(StandardizedSchedule standardizedSchedule);

	protected abstract MapperS<BigDecimal> initialMarginRequirement(StandardizedSchedule standardizedSchedule);

	public static class GetGrossInitialMarginFromStandardizedScheduleDefault extends GetGrossInitialMarginFromStandardizedSchedule {
		@Override
		protected Money.MoneyBuilder doEvaluate(StandardizedSchedule standardizedSchedule) {
			Money.MoneyBuilder grossInitialMargin = Money.builder();
			return assignOutput(grossInitialMargin, standardizedSchedule);
		}
		
		protected Money.MoneyBuilder assignOutput(Money.MoneyBuilder grossInitialMargin, StandardizedSchedule standardizedSchedule) {
			grossInitialMargin
				.setValue(MapperMaths.<BigDecimal, BigDecimal, BigDecimal>multiply(MapperMaths.<BigDecimal, BigDecimal, BigDecimal>multiply(MapperS.of(standardizedSchedule).<BigDecimal>map("getNotional", _standardizedSchedule -> _standardizedSchedule.getNotional()), initialMarginRequirement(standardizedSchedule)), MapperS.of(new BigDecimal("0.01"))).get());
			
			grossInitialMargin
				.getOrCreateUnit()
				.setCurrencyValue(MapperS.of(standardizedSchedule).<String>map("getNotionalCurrency", _standardizedSchedule -> _standardizedSchedule.getNotionalCurrency()).get());
			
			return Optional.ofNullable(grossInitialMargin)
				.map(o -> o.prune())
				.orElse(null);
		}
		
		@Override
		protected MapperS<BigDecimal> initialMarginRequirement(StandardizedSchedule standardizedSchedule) {
			return MapperS.of(getStandardizedScheduleMarginRate.evaluate(MapperS.of(standardizedSchedule).<StandardizedScheduleAssetClassEnum>map("getAssetClass", _standardizedSchedule -> _standardizedSchedule.getAssetClass()).get(), MapperS.of(standardizedSchedule).<BigDecimal>map("getDurationInYears", _standardizedSchedule -> _standardizedSchedule.getDurationInYears()).get()));
		}
	}
}
